packages/coinjoin/src/client/CoinjoinRound.ts

```diff
@@ -140,11 +141,31 @@ export class CoinjoinRound extends EventEmitter {
     }
 
     async onPhaseChange(changed: Round) {
-        // if round is currently locked interrupt running process
         if (this.lock) {
-            this.options.log(`Aborting round ${this.id}`);
-            this.lock.abort();
-            await this.lock.promise;
+            // if round is currently locked and phase was changed in expected order
+            // try to interrupt running process and start processing new phase
+            // but give current process some time to cool off
+            // example: http request is sent but response was not received yet and aborted
+            const shouldCoolOff = changed.phase === this.phase + 1;
+            const { promise, abort } = this.lock;
+            const unlock = () => {
+                this.options.log(`Aborting round ${this.id}`);
+                abort();
+                return promise;
+            };
+
+            if (!shouldCoolOff) {
+                await unlock();
+            } else {
+                this.options.log(
+                    `Waiting for round ${this.id} to cool off ${ROUND_PHASE_PROCESS_TIMEOUT}ms`,
+                );
+                // either process will finish gracefully or will be aborted
+                await Promise.race([
+                    promise,
+                    scheduleAction(unlock, { delay: ROUND_PHASE_PROCESS_TIMEOUT }),
+                ]);
```

Comment:
Does it matter that `unlock` in `scheduleAction` will be called even if `promise` resolves before that? Also, this will probably work the same way without the need of `Promise.race`. :arrow_down:  Do you see any pros or cons?
```suggestion
                await scheduleAction(() => promise, { timeout: ROUND_PHASE_PROCESS_TIMEOUT }).catch(
                    unlock
                );
```

Reply:
good point

Reply:
done, squashed